hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/TransactionManager.java

```diff
@@ -38,40 +43,131 @@ public class TransactionManager implements Serializable, AutoCloseable {
+
+  /**
+   * Uses the provided instant if present or else generates an instant time and executes an action that requires that instant time within a lock.
+   * @param providedInstantTime an optional instant time provided by the caller. If not provided, a new instant time will be generated.
+   * @param lastCompletedActionInstant optional input representing the last completed instant, used for logging purposes.
+   * @param instantTimeConsumingAction a function that takes the generated instant time and performs some action
+   * @return the result of the action
+   * @param <T> type of the result
+   */
+  public synchronized <T> T executeStateChangeWithInstant(Option<String> providedInstantTime, Option<HoodieInstant> lastCompletedActionInstant, Function<String, T> instantTimeConsumingAction) {
+    if (isLockRequired()) {
+      acquireLock();
+    }
+    String requestedInstant = providedInstantTime.orElseGet(() -> HoodieInstantTimeGenerator.createNewInstantTime(timeGenerator, 0L));
+    try {
+      if (lastCompletedActionInstant.isEmpty()) {
+        LOG.info("State change starting for {}", changeActionInstant);
+      } else {
+        LOG.info("State change starting for {} with latest completed action instant {}", changeActionInstant, lastCompletedActionInstant.get());
+      }
+      return instantTimeConsumingAction.apply(requestedInstant);
+    } finally {
+      if (isLockRequired()) {
+        releaseLock();
+        LOG.info("State change ended for {}", requestedInstant);
+      }
+    }
+  }
+
+  public void beginStateChange() {
+    beginStateChange(Option.empty(), Option.empty());
+  }
+
+  public synchronized void beginStateChange(Option<HoodieInstant> changeActionInstant,
+                                            Option<HoodieInstant> lastCompletedActionInstant) {
     if (isLockRequired) {
       LOG.info("State change starting for {} with latest completed action instant {}",
           changeActionInstant, lastCompletedActionInstant);
-      lockManager.lock();
+      acquireLock();
       reset(this.changeActionInstant, changeActionInstant, lastCompletedActionInstant);
       LOG.info("State change started for {} with latest completed action instant {}",
           changeActionInstant, lastCompletedActionInstant);
     }
   }
 
-  public void endStateChange(Option<HoodieInstant> changeActionInstant) {
+  public void endStateChange() {
+    endStateChange(Option.empty());
+  }
+
+  public synchronized void endStateChange(Option<HoodieInstant> changeActionInstant) {
     if (isLockRequired) {
       LOG.info("State change ending for action instant {}", changeActionInstant);
       if (reset(changeActionInstant, Option.empty(), Option.empty())) {
-        lockManager.unlock();
+        releaseLock();
         LOG.info("State change ended for action instant {}", changeActionInstant);
       }
     }
   }
 
+  private void acquireLock() {
+    if (hasLock) {
+      LOG.debug("Lock already acquired, skipping lock acquisition.");
```

Comment:
okay this is not a safe thing to do in general.. If the underlying lock is distributed, we want actually to attempt locking again and throw an exception instead.. rn, we will safely assume we have the lock.. 

the lock manager needs to ensure re-entrancy if thats required. This is the issue for the 1 failing test

Reply:
I think deltastreamer has, by virtue of sharing write clients.. I am going to change stuff such that it works with distributed locks. with both concurrent access and separate processes..

Reply:
Short story:  existing test is kind of broken. but making `beginStateChange`/ `endStateChange` will cause a deadlock.. second writer will wait for the lock within beginStateChange.. and the first writer cant release it. 

So we need to keep these methods simple.. and no shared state.